### ISSUE TITLE
5 times faster over all and massive reduction in the amount of code.

### ISSUE THREAD

**Issue by myearwood1** [VFPX/GoFish#142]
Method to massively reduce code and improve performance 500%

Incorporate a wrapper for vfp2c32 adirex to scan files directly to a cursor instead of an array (while accessing the entire VFP Search Path) and permit SQL commands against the list of files instead of slow high-level looping code.

Since so many programmers think premature optimization is bad, they do not even know how to optimize. Donald Knuth said a 12% improvement would be accepted by any industry but programmers.

My experience with this community is that even 5 times faster is disregarded. Such dismissal of skillful programmers harms the community instead of improving it. Allowing the continued presence of a person that LIBELLED me is insulting.

MySearch.prg
a=seconds()

*get the files
IF DirX('c_TheFiles','VFPPATH*.*','',.T.) > 0

	*Which files contain "REPORT FORM"?
	select occurs('REPORT FORM',upper(filetostr(alltrim(fullname)))) as nHowMany,*  ;
	from ;
		c_TheFiles ;
	where ;
		fileext = 'PRG' and upper(filename) # 'MYSEARCH.PRG' ;
	having ;
		nHowMany > 0 ;
	order by ;
		fileext ;
	INTO CURSOR ;
		JUSTTHESE ;
	NOFILTER

endif

use in (select("c_thefiles"))

?seconds()-m.a

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

**Comment by myearwood1** [comment]
It seems I am being deliberately kept from any input into this product. Lutz went off and started his own derivative. I showed this programming so you guys could take DIRX and stop using adir and objects when FoxPro is wonderfully capable with cursors.

This deliberate exclusion is being reported to GitHub

**Comment by myearwood1** [comment]
Based on Doug's input about using different flls, I am making some changes. Drew Speedie and Tom Demay made a program (x2setlib) which loads a dll if it can and if it is not loaded. That program should be 2 pieces, one that detects if the fll is loaded and another that does the loading.  By that you can attempt to load the fll. If it succeeds, we will know it was loaded. If it is already loaded, it will do nothing. That way the place in your code where the fll is loaded, can successfully release it. Once loaded, subsequent calls to load will not load.

I have made my own version of DeleteFileEx from vfp2c32.fll which would allow you to delete all GF Janitor files with one command. 

select ;
	DeleteFileExTemp(alltrim(filename),.T.,2) as DeleteNumber ;
	from c_Temp ;
	having DeleteNumber <> 0 ;
	into cursor c_CouldNotDeleteNumber

I usually have a .prg load and defer to the fll/dll function. Those will all load and unload the fll unless it was loaded before. In my case, I might load the FLL when the app starts.

LOCAL llLoaded
llLoaded=x2isLib('vfp2c32.fll')
if not m.llLoaded
	llLoaded = x2setlib('vfp2c32.fll')
	if not m.llLoaded
		wait window 'Could not load....'
		return .f.
	endif
endif

DeleteFileEx,prg can check that, defer to the FLL and unload the fll if DeleteFileEx.prg loaded it. For performance, load the fll before calling the sql simply by

llLoaded = x2setlib('vfp2c32.fll')
run the sql
if llLoaded
  release library vfp2c32.fll
endif

That will be all you need to do in GFJanitor

**Comment by Jimrnelson** [comment]
@myearwood1, @DougHennig 

I am still confused by this issue (competing FLLs) and need clarification, particularly because this is something I cannot test myself and also because it absolutely has to work in all cases.

From what I have read here, it's seems straightforward enough if there's just one FLL involved, the one used by DIRX.

However, I do not understand how it is to be handled if the other (same named) FLL has already been loaded,  I do not understand how it is possible to sense this nor how the sample code (in the previous post) handles this.

In the case where the other FLL is already loaded, it seems to me that somehow or other the "other" FLL must be temporarily disabled, the "correct" FLL enabled, and thereafter the two reversed.  I need specific direction on how to do this or I will not be able to proceed,

---

While DIRX looks like a significant improvement, I believe the real improvement will come from this:

> directly to a cursor instead of an array (while accessing the entire VFP Search Path) and permit SQL commands against the list of files instead of slow high-level looping code.

And this can be accomplished with or without DIRX.

---

As for the janitor, as of today GF defaults to only running it once a week, so it has dropped to a very low priority.  There have been significant improvements in the last few weeks in the management of the files that the janitor takes care so that this is likely a moot issue anyway.

**Comment by Jimrnelson** [comment]
@DougHennig, @myearwood1 

Whew!  That's good to hear and nothing could be easier.

**Comment by myearwood1** [comment]
In VFP2c32 there is a function that extends ADIR, but ADIR is IMO always the wrong approach. DirX is a wrapper around ADirEx, giving the ability to scan the vfppath as well as accepting multiple file specs. DirX already has a switch to shut down VFP2C32 if it loads it.

However, deleting the files can be done with the Windows DeleteFile API itself without needing VFP2c32.

DeleteFile.prg
LPARAMETERS m.lpFileName
DECLARE INTEGER DeleteFile IN kernel32 STRING @lpFileName
RETURN DeleteFile(@m.lpFileName)

select DeleteFile(alltrim(filename)) as nDeleted ;
from some cursor ;
into cursor NotDeletedCursor nofilter

I know of no faster way to get a cursor of files than VFP2c32's C++ library. I made the DirX wrapper to hide some of the complexity of calling vfp2c32.ADirEx

I tried 50,000 files earlier and it deleted them all in 6.5 seconds. The ADIR approach crashes. Building the cursor in foxpro code is definitely slower but you could do it.

dirx("c_temp",addbs(sys(2023))+"*.tmp")
select DeleteFile(alltrim(filename)) as nDeleted ;
	from c_Temp ;
	into cursor c_Deleted nofilter

**Comment by myearwood1** [comment]
and you do not need to load vfp2c32 until the search begins, so theoretically, if a developer had V1 loaded, you might have to unload that one, load v2 and unload v2 and then reload v1.

**Comment by Jimrnelson** [comment]
@myearwood1 , @DougHennig 

> and you do not need to load vfp2c32 until the search begins, so theoretically, if a developer had V1 loaded, you might have to unload that one, load v2 and unload v2 and then reload v1.

See Doug's earlier reply in this thread.  He says that there is no need to worry about the distinct versions.

He says this:

> * VFP2C32 (any version) is loaded when GF starts: GF doesn't need to do anything.

**Comment by myearwood1** [comment]
Which I think means no version of VFP2C32 was already loaded. I mean if some developer had the old version loaded and there was some conflict, possibly a missing function, you would have to unload the old version, load the new, unload the new, reload the old, at worst. I have never tried 2 FLLs with conflicting function names. I also cannot say off the top of my head whether in the case of vfp2c32 having the same function name as a Windows API, which one will take effect.

**Comment by Jimrnelson** [comment]
@myearwood1 , @DougHennig 

If there is any possibility of a conflict here (Doug says there isn't), I am not the one who can write the code to load/unload the various each of them.  Nor could I test it.

I am getting more and more leary of using DirX because of this.  There are only two possible use cases:

[1]  When doing a folder search in GF.  But I have already identified that this is only a very small percentage of the total time doing a single GF search, so I see no real advantage to optimizing it.

[2]  When used by the Janitor at the end of the session.  There were problems in GF 6 in that the janitor did not always work correctly and now in GF 7 there are many many fewer files that need the janitor's attention.  Again, I do not see much of an upside.

Thus, unless it there is a clear and easy procedure for handling the case of the "other" VFP2C32 file being already loaded, I am going to leave this as is.

That being said, I think that the real opportunity of optimizing is in Mike's original statement:

> scan files directly to a cursor instead of an array (while accessing the entire VFP Search Path) and permit SQL commands against the list of files instead of slow high-level looping code.

**Comment by DougHennig** [comment]
The only conflict I'm aware of is the InitVFP2C32 function I mentioned, but without going throught the two FLLs functions by function can't guarantee that. You could always do something like this:

if 'VFP2C32.FLL' $ UPPER(SET('LIBRARY'))
    release library VFP2C32
endif
set library to VFP2C32

That would ensure whatever version is loaded is unloaded and then the correct version (V2) is loaded.

**Comment by Jimrnelson** [comment]
> The only conflict I'm aware of is the InitVFP2C32 function I mentioned, but without going throught the two FLLs functions by function can't guarantee that. You could always do something like this:
> 
> if 'VFP2C32.FLL' $ UPPER(SET('LIBRARY')) release library VFP2C32 endif set library to VFP2C32
> 
> That would ensure whatever version is loaded is unloaded and then the correct version (V2) is loaded.

Yes that would get the correct version loaded but not necessarily take care of reloading the original one.

Beyond that, it seems that `release library` has some problem with libraries that have spaces in the path (as mine does for FoxTools).  Thus `release library all` does not release foxtools for me.  Similarly, despite what the help file says, `release library all` does not release FoxTools.  It seems like one cannot reliably remove a library.

**Comment by myearwood1** [comment]
If the other version is loaded by the developer, we have to respect that environment by preserving it, changing to the new one and then restoring it. That sounds like it's all on me if I want to offer DirX to you guys.

It's less of an issue for me now too. This app has been accumulating temp files for years!
I will test the various scenarios. I wanted to try Carlos' code. Not every app has to contend with possible conflicting FLLs/APIs but GoFish could.

As I was trying to explain to Christian who makes VFP2C32, libraries bug me because there is always the chance two libraries will have some routine with the same name. In a DLL, I can alias the function, but not so with an FLL.

If you're using adir to recurse the folders, you're risking is crashing. DirX can recurse all of the drive.

**Comment by Jimrnelson** [comment]
> This app has been accumulating temp files for years!

Yes, GF, like Thor, PEMEditor and other tools I am aware of, create files in SYS(2023) and makes no attempt to delete them later.  No doubt there are a lot of such small files.

> If you're using adir to recurse the folders, you're risking is crashing.

GF has always used ADIR to recurse folders and to the best of my knowledge there has never been a report of a crash.

**Comment by myearwood1** [comment]
That's what they say here all the time. FoxPro 6 then it blows up. I had to stress the need to increase mvcount from the default. Been that way for 20 years. Just recently out of the blue started giving error message Too Many Variables. Remember MS increased ADIR to a ridiculous level because it was too small. We lose speed by walking the array, because this is not c++.

**Comment by myearwood1** [comment]
Or one I just worked on. 2 records with the same date/time. Fox is unable to give enough time resolution. You're only safe until you're not.

**Comment by myearwood1** [comment]
Fellows, I did find a little snag to be wary of. I am a huge fan of Drew Speedie's load and defer (proxy) method. If turns out there can be a problem not related to other versions of vfp2c32, but between vfp2c32 and the Windows API.

If I call GetFileAttributes('somefile'), my getfileattributes.prg fires, declares the windows API getfileattributes and calls it directly. Thereafter, every call to getfileattributes in my code calls the windows API.

If I first load vfp2c32, it has a getfileattributes which does not behave identically to the Windows API. That means my getfileattributes.prg never fires, and so subsequent calls to getfileattributes crash. That's only if the vfp2c32 is loaded first.

If vfp2c32 is not loaded, my getfilesattributes proxy loads the DLL and thereafter not even loading vfp2c32 affects the system. The DLL takes precedence over the FLL and both take precedence over the FoxPro .prg.

None of this matters specifically to removing the temp files, but it is interesting.

If you want to delete tempfiles and with a log of error numbers, that is an easy thing. If you want error messages about the failure to delete, that needs a little more help. Do you care if the temp files fail to delete or not? Or do we just want the attempt to be made without crashing like ADIR can?

**Comment by Jimrnelson** [comment]
Mike --

Um, I think this (deleting temp files) should be a separate issue.  This one is getting crowded.

As it is I am not quite sure what you are referring to when you say "temp files".  

You mean the work that the Janitor does?  If so, you should know that the janitor is erasing "old" search history files and some few other junk files in the same folder.  It will never just do a blind deletion of all files in the folder.

Right now, this is handled by a single call to ADIR and the only possible reason for replacing it with DIRX would be to prevent the highly unusual case of ADIR failing because of too many files.  Yes, you reported the problem, but the problem only occurred because of errors (since corrected) that let your folder grow tremendously.

I am not at all convinced that this is worth pursuing.

---

Addenda:  There are two ways for handling search history (and I have no idea which is prevalent)

- Users manage search history manually, by saving searches they want to keep and using the history form to delete those that are no longer needed.  This avoids the janitor altogether.
- Users have all searches saved automatically and let the janitor clean up automatically.  (It runs the first time you close GF each day, although the frequency is now an option). 

It is my belief that the changes I have made in GF 7 have so minimized the problem (by reducing the number of files created in that folder by a factor of at least 10) that no further changes to the janitor are needed.

**Comment by myearwood1** [comment]
If the janitor is doing better then good. We have since had each instance
of our app create its own temp folder and a config.fpw. then I'm still
erasing .tmp files at launch.

Foxpro must be creating tmp files as gofish runs. Those being put in the
same folder as all other windows temp files is a concern.

**Comment by myearwood1** [comment]
As to what is worth pursuing, too many programmers ignore what to me are obvious improvements. I expect there are shortfalls in what is delivered and that is why I knew to look for and speed up the RegEx thing - which you said is 7 times faster.

**Comment by myearwood1** [comment]
@Jimrnelson you said 

> > The only conflict I'm aware of is the InitVFP2C32 function I mentioned, but without going throught the two FLLs functions by function can't guarantee that. You could always do something like this:
> > if 'VFP2C32.FLL' $ UPPER(SET('LIBRARY')) release library VFP2C32 endif set library to VFP2C32
> > That would ensure whatever version is loaded is unloaded and then the correct version (V2) is loaded.
> 
> Yes that would get the correct version loaded but not necessarily take care of reloading the original one.
> 
> Beyond that, it seems that `release library` has some problem with libraries that have spaces in the path (as mine does for FoxTools). Thus `release library all` does not release foxtools for me. Similarly, despite what the help file says, `release library all` does not release FoxTools. It seems like one cannot reliably remove a library.

Assuming you SET LIBRARY TO (HOME()+'foxtools.fll"), have you done RELEASE LIBRARY (HOME()+"foxtools.fll")?

To my knowledge that is the only way to unload a library with a path that has spaces.

**Comment by Jimrnelson** [comment]
@myearwood1 

Unfortunately,

    RELEASE LIBRARY (HOME()+"foxtools.fll")

has no effect.

Similarly, using 

    Set Library to vfp2c32.fll

without the `Additive` keyword does not remove foxtools.fll.

**Comment by myearwood1** [comment]
I think if cygwin's grep is used, none of this thread will be needed.
